feat: add hex scalar function

tshauck opened the pull request:
## Which issue does this PR close?

Related to https://github.com/apache/datafusion-comet/issues/341.

## Rationale for this change

I recently added `unhex` so this PR adds `hex`. It's another scalar function that isn't yet implemented in Comet.

I decided to do a native version here because datafusion has an `to_hex`, but it has different functionality. E.g. `to_hex(-1)` returns `ffffffffffffffff` in datafusion vs `hex(-1)` returning `FFFFFFFFFFFFFFFF` in spark. Spark `hex` also supports strings, but datafusion and postgres do not. I'm happy to start the discussion if it seems it could be merged upstream.

## What changes are included in this PR?

I added the `hex` scalar function.

## How are these changes tested?

Yes, I've added tests to the rust side as well as spark sql based tests in scala.

kazuyukitanimura commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,191 @@
+
+    match &args[0] {
+        ColumnarValue::Array(array) => match array.data_type() {
+            DataType::Int64 => {
```
Wondering if we need to support more types?

viirya replied:
Spark `Hex` expression only supports LongType, BinaryType, StringType.

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,46 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
   }
+  test("hex") {
+    val str_table = "string_hex_table"
+    withTable(str_table) {
```
Let's add more types and values, something like `makeRawTimeParquetFile`

tshauck replied:
Thanks, will have a look at `makeRawTimeParquetFile` and follow up.

kazuyukitanimura replied:
probably `makeParquetFileAllTypes` is a better example

tshauck replied:
Thanks for the feedback. I update the tests in https://github.com/apache/datafusion-comet/pull/449/commits/2bfcf25a6705ec47d857670c65dbba979ca10e23 to use `makeParquetFileAllTypes`. And with a couple of changes its tests passes w/ the ints, uint32 and uint64[1], and the _8 and _14 which I think are string/binary types.

When the #451 goes in, I'll look at updating the tests to add it too.

[1]: I'm a bit confused because it looks like every int crossing into rust does so as an i64, which means rather than properly return say `FF` for u8, I'm returning `FFFFFFFFFFFFFFFF`. I need to look into this further.

kazuyukitanimura replied:
@advancedxy got a good point. Spark may be casting ints to `i64`. Then we only need to handle `i64` in Rust as long as it passes the tests.

tshauck replied:
Yeah, I think that's the case. I'll pare down the rust code.

tshauck commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,191 @@
+fn hex_bytes_to_string(bytes: &[u8]) -> Result<String, std::fmt::Error> {
+    let mut hex_string = String::with_capacity(bytes.len() * 2);
+    for byte in bytes {
+        write!(&mut hex_string, "{:01X}", byte)?;
```
I'll look into `:02X`, but it seems to pass tests w/ `:01X`. Also on `hex_encode`, I wonder if the hash functions should go in `scalar_funcs/hash.rs` or something, then `hex_encode` could go there? Personal preference, but I think it'd be nice if `scalar_funcs` was refactored so it just had `create_comet_physical_fun` and the individual function impls went in `scalar_funcs`.

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,26 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+          // ints
+          checkSparkAnswerAndOperator("SELECT hex(_2), hex(_3), hex(_4), hex(_5) FROM tbl")
+
+          // uints, uint8 and uint16 not working yet
```
I remember unsigned ints currently do not work in other places https://github.com/apache/datafusion-comet/blob/main/spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala#L97

I think it is ok to leave it as TODO

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,26 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+          checkSparkAnswerAndOperator("SELECT hex(_11), hex(_12) FROM tbl")
+
+          // strings, binary
+          checkSparkAnswerAndOperator("SELECT hex(_8), hex(_14) FROM tbl")
```
What about other types like _1, _6, _7, _13, _15+?

tshauck replied:
1caa6fd adds the other columns... `_13` also fails failure.

tshauck replied:
I updated dictionary handling, so I've added _13 back in. I think only the smaller uints are omitted now.

tshauck commented:
@kazuyukitanimura @advancedxy ... re-requesting reviews from you two, please. I've updated the code to support dictionaries and removed some of the finer int types. Barring something happening on intel mac, the tests look to pass here: https://github.com/tshauck/arrow-datafusion-comet/actions/runs/9183637020.... thanks

andygrove commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,371 @@
+
+                Ok(ColumnarValue::Array(Arc::new(string_array)))
+            }
+            DataType::Dictionary(_, value_type) if matches!(**value_type, DataType::Int64) => {
```
I also ran into issues with this in handling cast operations and ended up flattening the dictionary type first, but just for cast expressions. I agree that we need to look at this and see if there is a more systematic approach we can use.

kazuyukitanimura replied:
> Do you have any concrete examples of how this works by any chance?
Hmm, it's been a while, I cannot find right away...

BTW my comment above is not a blocker. Since this PR already implemented it, we can follow up separately

kazuyukitanimura replied:
BTW now I remember `unpack_dictionary_type()` unpacks early for primitive types, so like the one for Int64 can be omitted.

tshauck commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,20 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+        withParquetTable(path.toString, "tbl") {
+          // _9 and _10 (uint8 and uint16) not supported
+          checkSparkAnswerAndOperator(
+            "SELECT hex(_1), hex(_2), hex(_3), hex(_4), hex(_5), hex(_6), hex(_7), hex(_8), hex(_11), hex(_12), hex(_13), hex(_14), hex(_15), hex(_16), hex(_17), hex(_18), hex(_19), hex(_20) FROM tbl")
```
Thanks, I agree it's best to test since they're there. I just genuinely thought it wasn't supported given the test coverage, the errors on the other functions, etc. I'll have a look at adding some tests...

tshauck replied:
I'm realizing that I'm not going to have time next week and didn't expect this PR to take this long, so I've removed the scalar handling for now and hopefully can follow and example in the future. https://github.com/apache/datafusion-comet/pull/449/commits/88bdcde1babcd75cb52e18164e1c945bbbbbf857

tshauck replied:
I initially tried and got test failures because the native function isn't being recognized w/ scalars. The same issue I mentioned w/ `trim` above.

kazuyukitanimura replied:
Sounds like QueryPlanSerde is failing the case match. 
Wondering if you already have tried `"spark.sql.optimizer.excludedRules" -> "org.apache.spark.sql.catalyst.optimizer.ConstantFolding"` ?

Also right now, since the scalar code was removed from rust, it will fail if the native code happen to find scalar values...

tshauck replied:
Thanks, yes, I did also try disabling ConstantFolding to no avail.

viirya replied:
`hex(literal)` should be evaluated to literal early in Spark optimizer. I wonder what test failure you saw?

viirya commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,296 @@
+                    .map(|v| v.map(hex_int64))
+                    .collect::<Vec<_>>();
+
+                let keys = dict.keys().clone();
+                let mut new_keys = Vec::with_capacity(values.len());
+
+                for key in keys.iter() {
+                    let key = key.map(|k| values[k as usize].clone()).unwrap_or(None);
+                    new_keys.push(key);
+                }
+
+                let string_array_values = StringArray::from(new_keys);
```
Hmm, why we don't re-create a new dictionary array of string?

andygrove replied:
We have the same issue currently in cast from string to other types. 

@viirya  Do we have an example somewhere of converting a dictionary array without unpacking?

viirya replied:
For this case, we can call

```
dictionary.with_values(Arc::new(values));
```

to construct a new dictionary array with new values.

tshauck replied:
@viirya, thanks, where I'm a bit unclear is how to have the function return type also be a dictionary. The data type for the hex expression seems to be a Utf8, so I get `org.apache.comet.CometNativeException: Arrow error: Invalid argument error: column types must match schema types, expected Utf8 but found Dictionary(Int32, Utf8)` after making the update to use `with_values`.

andygrove replied:
@viirya Am I correct in understanding that this PR is functionally correct but just not as efficient as possible? Perhaps we could consider having a follow-up issue to optimize this to rewrite the dictionary? It seems that we don't have a full example of dictionary rewrite for contributors to follow.

viirya replied:
@andygrove Yes. It's not efficient but should be correct.

viirya commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,20 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+        }
+      }
+    }
+  }
   test("unhex") {
```
style:
```suggestion
  }
  
  test("unhex") {
```

viirya reviewed (changes requested):
I left a few comments including ones that need to be addressed before merging this, e.g. https://github.com/apache/datafusion-comet/pull/449#discussion_r1617541513.

tshauck commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -1038,6 +1038,20 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       }
     }
   }
+  test("hex") {
```
[e7062c6](https://github.com/apache/datafusion-comet/pull/449/commits/e7062c61a42ec254ec5db203845ec66e21a56bf9)

tshauck commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,296 @@
+                Ok(ColumnarValue::Array(Arc::new(hexed_array)))
+            }
+            DataType::Utf8 | DataType::LargeUtf8 => {
+                let array = as_string_array(array);
```
129f00ab1379f16ac6f3a68bb1dc73dc1fcebeb8

viirya commented on core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:
```diff
@@ -0,0 +1,306 @@
+                let string_array_values = StringArray::from(new_keys);
+                Ok(ColumnarValue::Array(Arc::new(string_array_values)))
+            }
+            DataType::Dictionary(_, value_type) if matches!(**value_type, DataType::Binary) => {
```
This part of dictionary arrays handling can be rewritten to reduce duplicate actually. It can be follow-up though.

andygrove replied:
I filed https://github.com/apache/datafusion-comet/issues/504

andygrove reviewed (approved):
Thanks again @tshauck
